Add SCSR parsing to RDKit (#8147)
rdkit/rdkit: Docs/Book/RDKit_Book.rst

been generated, it is used to set multiple bits based on different atom and bond
type definitions.

> Self-Contained Structure Representations (SCSR) for Macromolecules

Reviewer: Thanks for writing this. It's really useful

*******************

The SCSR support tentative added to RDKit follows the description in the BIOVIA document “biovia_ctfileformats_2020.pdf” available from 
“CTfile Formats - Dassault Systèmes”.  That document does not provide any real detail for that format, and contains one example file.

In addition, Biovia Draw supports reading and writing this format.  As much as possible, the RDKit support allows the functionality

> supported by Biovia Draw.  One exception is the RDkit treatment of hydrogen bonds in SCSR files/blocks (vide infra).

Reviewer: ```suggestion
supported by Biovia Draw.  One exception is the RDKit treatment of hydrogen bonds in SCSR files/blocks (vide infra).
```

Author: ok

Representation
==============

Main mol
--------
An SCSR file contains a mol with a CTAB in v3000 format.   That CTAB can contain elemental atoms and macro atoms corresponding to amino acids (AA), RNA, and DNA elements.  
The RNA and DNA elements are represented by three parts – a phosphate group, a sugar, and a base.  

> Each atom line in the CAT can refer to an elemental atom or a macro atom.  Macro atom lines have a text description of the macro name and must have a CLASS and an

Reviewer: ```suggestion
Each atom line in the CTAB can refer to an elemental atom or a macro atom.  Macro atom lines have a text description of the macro name and must have a CLASS and an
```

Author: ok

ATTCHORD attributed.  They can also have an optional SEQID attribute.  

> According to the Biovia doc, the CLASS attribute must have a value that is one of:  AA, dAA, ScsrSupportInRDKitDNA, RNA, SUGAR, BASE, PHOSPHATE, LINKER, CHEM, LGR,P MODAA, MODdAA, MODDNA, MODRNA, XLINKAA, XLINKdAA, XLINKDNA, XLINKRNA.   

Reviewer: ```suggestion
According to the Biovia doc, the CLASS attribute must have a value that is one of:  AA, dAA, DNA, RNA, SUGAR, BASE, PHOSPHATE, LINKER, CHEM, LGRP,  MODAA, MODdAA, MODDNA, MODRNA, XLINKAA, XLINKdAA, XLINKDNA, XLINKRNA.   
```

Author: OK

> This implementation in RDKit does NOT enforce the allowed list.

Reviewer: I think it does enforce the allowed list.

Author: SO I took a look.  The class in an atom definition is NOT checked.  There is an example in Code/GraphMol/FileParsers/file_parsers_catch.cpp that has a class of "foo".

The class in an SGROUP is checked.   In SCSR block parsing, the atom line will have a class which is not checked against the allowed list, but this must also appear in an SGROUP and that is checked, so the SCSR treatment DOES enforce the allowed list.  

Thanks for catching this.  I have changed the doc to reflect this.

The SEQID is a sequential integer and is ignored by this treatment.  Typically, the three parts of an RNA or DNA element have the same SEQID.

The ATTCHORD attribute must have a specification for each bond that comes from the macro atom.   The specification is contained between parentheses, and the first 
character is a integer that indicates the number of items that follow.  Each connection is specified with the atom ID (1 offset) for the connected atom in the main
CTAB, and a string that indicates the attachment point for this macro atom.   The attachment point string is always two characters.  
The first indicates the order, and is a capital letter starting at A.  The second char is one of “l” (left), “r” (right), “x” (cross connections – e.g. for cysteine).  
In this implementation, the second character can also be “h” for hydrogen bond.   

> Thus, the attach connections are almost always in the list: “Al”, “Br”, “Cx” or “Ch”.   For example:
> 
>     # ATTCHORD=(6 15 Br 13 Al 21 Cx)

Reviewer: It's not clear to me why this starts with a `#`.
In order to get these to render as code samples, which I think is what you want, you need to end the previous line with `::` (yes, RST is irritating).
So this becomes:
```suggestion
Thus, the attach connections are almost always in the list: “Al”, “Br”, “Cx” or “Ch”.   For example::

    ATTCHORD=(6 15 Br 13 Al 21 Cx)
```

Please update all the other examples too.

Author: OK

Templates
---------

Template Header
^^^^^^^^^^^^^^^

In addition to the CTAB for the main molecule, each macro atom is detailed automatically in a TEMPLATE.  The TEMPLATES are numbered and appear between a 
BEGIN TEMPLATES line and an END TEMPLATES line. Each template starts with a TEMPLATE line that indicates the template number (1 to n), the template Class and 
Name, and the NATREPLACE attribute.

> The Class and Name consists of three (or more) parts separated by forward slashes.  The first part is the CLASS,  and must match the CLASS attribute in 

Reviewer: ```suggestion
The Class and Name consists of three (or more) parts separated by forward slashes.  The first part is the CLASS, and must match the CLASS attribute in 
```

Author: OK

“Ch” will typically have two or three distinct SAP attributes.  The order of the SAP attributes for hydrogen bonds (“Ch”) matters, and should go from the H-bond 
atom nearest to the connecting “Al” atom to the most distant.  Hydrogen bonds do not have a leaving group, so the second ID of the designation is “0”.

Example:

    # M  V30 2 SUP 2 ATOMS=(8 1 2 3 4 5 6 7 8) XBONDS=(1 7)-
    M  V30 LABEL=U CLASS=BASE SAP=(3 4 9 Al) SAP=(3 8 0 Ch) SAP=(3 6 0 Ch) SAP=(3 7 0 Ch)-
    NATREPLACE=BASE/U

In addition to the main SGROUP, there will be an SGROUP that identifies each leaving group.  Most leaving groups have one atom, but they could have multiple 
atoms.  The leaving group SGROUPS have ATOMS, XBONDS, LABEL, and CLASS attributes.   The XBONDS attribute is ignored in this treatment.   LABEL must be 
“LGRP” (leaving group).

Example:

    # M  V30 1 SUP 1 ATOMS=(1 11) XBONDS=(1 11) LABEL=H CLASS=LGRP

Parsing SCSR files and text blocks
==================================

The following calls parse the SCSR mol or block produces an SCSRMol:

  * SCSRMolFromScsrDataStream  (stream to SCSRMol)
  * SCSRMolFromScsrBlock  (text block to SCSRMol)

>   * SCSRMolFromScsrFile  (file to SCSRMol)

Reviewer: This block documents functionality that's not in the public API, please remove

> Full Example of SCST files:
> ===========================
> 
> Here are two examples of SCST files:

Reviewer: ```suggestion
Full Example of SCSR files:
===========================

Here are two examples of SCSR files:
```

Author: OK

> `DnaTest3 </Docs/Book/data/DnaTest3.mol>`_.
> 
> `CrossLink.mol </Docs/Book/data/CrossLink.mol>`_.

Reviewer: These links do not work in the final documentation.

You either need to figure out how to make Sphinx copy the files and what the format of the links should be or include links to where the files will be in Github, like this:
(i.e. ``)
```suggestion
`DnaTest3.mol <https://github.com/rdkit/rdkit/blob/master/Docs/Book/data/DnaTest3.mol>`_

`CrossLink.mol <https://github.com/rdkit/rdkit/blob/master/Docs/Book/data/CrossLink.mol>`_
```

Author: I made these changes as suggested.  That link does not work for me either because those files are not in master yet

  
Expansion to Full Atomistic Form

> ===============================

Reviewer: This block should be adapted to document the functions which are actually in the public API.

Author: reworked the doc to reflect the actual calls that are available

Reviewer: > reworked the doc to reflect the actual calls that are available

Great, please let me know when you have pushed the changes and I will take a look.

It is possible to convert an SCSRMol into a full atomistic representation as a RWMol.   That conversion is made with a call to MolFromSCSRMol.  This routine 
takes two parameters:  the SCSRMol to convert and a MolFromScsrParams parameter.
MolFromScsrParams has two properties at this time – a ScsrTemplateNames parameter and a Boolean ”includeLeavingGroups”.  

    * The “ScsrTemplateNames” parameter controls how the Sgoups of the expanded file are generated, and must be one of: ScsrTemplateNamesAsEntered, 
    ScsrTemplateNamesUseFirstName, or ScsrTemplateNamesUseLastName.   If ScsrTemplateNamesAsEntered is specified, the name as referenced in the main 
    SCSR CTAB will be used.  

>     * The ”includeLeavingGroups” parameter control whether leaving groups that are not replaced in the main CTAB are included in the resulting atomistic file.   

Reviewer: ```suggestion
    * The ”includeLeavingGroups” parameter controls whether leaving groups that are not replaced in the main CTAB are included in the resulting atomistic file.   
```

Author: ok
